### ISSUE TITLE
Undefined behavior in morton_split

### ISSUE THREAD

**Issue by kostinio** [madmann91/bvh#40]
Hello!
There is an undefined behavior in file bvh/include/bvh/morton.hpp, at this function
```
template <typename Morton>
Morton morton_split(Morton x) {
    constexpr size_t log_bits = round_up_log2(sizeof(Morton) * CHAR_BIT);
    auto mask = Morton(-1);
    for (size_t i = log_bits, n = 1 << log_bits; i > 0; --i, n >>= 1) {
        mask = (mask | (mask << n)) & ~(mask << (n / 2));
        x = (x | (x << n)) & mask;
    }
    return x;
}
```
`mask << n` and `x << n` is a shift of n-bit operand on n bits. This is an undefined behavior, because of

> if the value of the right operand is negative or is greater or equal to the number of bits in the promoted left operand, the behavior is undefined.
(https://en.cppreference.com/w/cpp/language/operator_arithmetic)

I found it when i tried to run this code with a clang. All my morton codes were 0.

**Comment by madmann91** [comment]
I think the first iteration of the loop can be peeled off manually to solve that issue

**Comment by madmann91** [comment]
Can you try the fix suggested in #41 ? To test it, just check out the branch named `morton_fix`.

**Comment by kostinio** [comment]
I tried, it works fine.
